lib/kafka/async_producer.rb

```diff
@@ -288,6 +280,29 @@ def do_loop
         retry
       end
 
+      def shutdown
+        # Deliver any pending messages first.
+        @producer.deliver_messages
+      rescue Error => e
+        @logger.error("Failed to deliver messages during shutdown: #{e.message}")
+
+        @instrumenter.instrument("drop_messages.async_producer", {
+          message_count: @producer.buffer_size + @queue.size,
+        })
+
+        if @finally
+          @queue.close
```

Comment:
[q] Wouldn't we also want to close the queue in the non-exceptional case?

Reply:
We close it in `#shutdown`.